tests/tests/Data/RawCodeCoverageDataTest.php

```diff
@@ -264,295 +262,6 @@ public function testRemoveCoverageDataForLines(): void
         $this->assertEquals($expectedFilterResult, $dataObject->lineCoverage());
     }
 
-    public function testUseStatementsAreUncovered(): void
-    {
-        $file = TEST_FILES_PATH . 'source_with_use_statements.php';
-
-        $this->assertEquals(
-            [
-                12,
-                14,
-                16,
-                18,
-            ],
-            array_keys(RawCodeCoverageData::fromUncoveredFile($file, new ParsingFileAnalyser(true, true))->lineCoverage()[$file])
-        );
-    }
-
-    public function testEmptyClassesAreUncovered(): void
-    {
-        $file = TEST_FILES_PATH . 'source_with_empty_class.php';
-
-        $this->assertEquals(
-            [
-                12,
-            ],
-            array_keys(RawCodeCoverageData::fromUncoveredFile($file, new ParsingFileAnalyser(true, true))->lineCoverage()[$file])
-        );
-    }
-
-    public function testInterfacesAreUncovered(): void
-    {
-        $file = TEST_FILES_PATH . 'source_with_interface.php';
-
-        $this->assertEquals(
-            [
-                7,
-                9,
-                11,
-                13,
-            ],
-            array_keys(RawCodeCoverageData::fromUncoveredFile($file, new ParsingFileAnalyser(true, true))->lineCoverage()[$file])
-        );
-    }
-
-    public function testInlineCommentsKeepTheLine(): void
-    {
-        $file = TEST_FILES_PATH . 'source_with_oneline_annotations.php';
-
-        $this->assertEquals(
-            [
-                13,
-                19,
-                22,
-                26,
-                29,
-                31,
-                32,
-                33,
-                35,
-                40,
-            ],
-            array_keys(RawCodeCoverageData::fromUncoveredFile($file, new ParsingFileAnalyser(true, true))->lineCoverage()[$file])
-        );
-    }
-
-    public function testHeavyIndentationIsHandledCorrectly(): void
```

Comment:
This test is very hardly crafted test, please keep it. Unit tests != functional tests, this test is more like a functional test and tests many specific combinations. `source_with_multiline_constant_return.php` test is also very important, at least keep these two, but I would prefer no single test deleted (if not migrated 1:1).

Reply:
I am aware of your previous work on `RawCodeCoverageDataTest`; I know the effort you put on it, also because I carefully crafted and reviewed it *two* times before you started contributing here: very time consuming and boring.

But all tests currently in `9.2` in `RawCodeCoverageDataTest` related to `ExecutableLinesFindingVisitor` must go:

1. The underlying logic has been completely changed: the data sets and the related comments just don't make sense anymore
2. The update needed to get the test pass marks all lines as executable, and wouldn't test how they are grouped by branch, so it wouldn't have any value as a regression test
3. You know how much of a PITA they currently are in `9.2` to both write and review, they are unbearable to maintain

Don't look at what we are losing, look at what we gained: this PR couldn't have been written without the precious knowledge and experience we got from working on it.

All the value we poured in our previous PR is still here: the (little) value you don't see anymore in raw data sets has been consumed in the form of a new architecture.

Reply:
I support the in place test annotation, but I the original data must be kept/tested 1:1. They tests many issues, syntetic or found during development and prevent https://github.com/sebastianbergmann/php-code-coverage/issues/942 which strictly requires executable lines to be no larger than maximum subset of real executable lines emit by coverage driver.

> All the value we poured in our previous PR is still here: the (little) value you don't see anymore in raw data sets has been consumed in the form of a new architecture.

That is what I am asking for, but why the original tests does not pass if I revert them?

Reply:
> why the original tests does not pass?

Because the logic changed.

> prevent #942 which strictly requires executable lines to be no larger than maximum subset of real executable lines emit by coverage driver.

The logic changed and this requirement is no longer valid.

This PR extends the green/red lines over all the AST node a driver passes.

Reply:
but this will massively return https://github.com/sebastianbergmann/php-code-coverage/issues/942 bug

I still strongly belive you should not emit more coverage points than is normally emit by xdebug.

Reply:
I will definitely test this PR. but please remind me why multiple green lines for for ex. multiline string are better than one?

I do not see any reason we should report/count more than one line for I source executable point/data. This is how Istanbul JS (and other coverage software) work.

In the description you mentioned (in real project) the coverage changed 93.65% to 96.11%. But what number is better? The first one has industry standard (+- xdebug/pcov/IstanbulJS) percentage and better informational value. If a string with 1M lines is added to a project, it must to skew the coverage by more than 1 executable line.

I understand the https://github.com/sebastianbergmann/php-code-coverage/issues/959 problem, I belive reasoned implied coverage points by executed branches can be a solution, but you should not report more lines than meaningfull (+- lines without opcache/non-costs expressions). So old tests should/must still pass and if you want to extend the coverage data for more lines, make it optional.

Also, as coded now this cannot be considered as a bugfix.

Reply:
Haven't looked at the PR in depth yet, but I think I would tend to agree that a hardcoded multiline string, or hardcoded multiline array should only count as 1 executable line - the assignment part is definitely executable code, but the contents thereafter are really just an opaque data blob once parsed - the choice of using '\n' or an actual newline in a string is irrelevant to which parts of the code are executed or not and it would be surprising to have code coverage numbers change if someone e.g. added an extra sentence into an email body.

Where a string has interpolation that calls into methods or an array has calculated values, then those particular lines I would expect to be identified as executable individually

Reply:
First things first, I managed to handle strings better, and now multiline strings have only 2 lines counted, the first and last ones; in between these two, no lines are marked anymore as executable.

But I urge you to focus on what really important this PR is finally achieving: reliability.

Of course I agree that only meaningful lines should produce CC, but what do we do when this cannot be done reliably? There's a scale of priorities we need keep in mind:

1. Non executed executable lines must be marked as red
2. Executed executable lines must be marked as green
3. CC reports must be deterministic and not influenced by the tools that produce or use them
4. Only meaningful lines should be reported

I'll do my best to achieve every listed point, but we *need* to sacrifice [4] *when* [1], [2] and [3] cannot be simultaneously satisfied.

The proposed solution achieve first three points *all* of the times and all four points *most* of the times, with a relative simpler code.

Reply:
You did not answer my question, what is the physical meaning (of emiting coverage for first and last string line, for empty lines, operator symbols, ...)?

In the PR description you reason this by:

> But as we know, line CC is convenient, quick and reliable only as a quantitative result.
Functions/Methods/Classes/Traits coverage has a much higher qualitative value, and this PR has very little effect on results. The difference in these cases are correct and expected: current 9.2.19 release has bugs that this PR fixes.

So basically you say line coverage has little imformation value (qualitative value). This is not true, as line coverage is normally the primary value reported by tools like CodeCov.

Also this PR has exactly oposite then little effect on results as 93.65% vs. 96.11% line coverage is huge change.

In the latest release we track executable lines well and #953/#959 is basically not even a bug, it is a feature request.

I belive you need to change the result of your branch analysis. We both know quite well when/where xdebug emit coverage points and we know when it does not (always or sometimes due opcache). The result I expect from branch analysis PR is an update of the original executable analysis to analyse two kinds of executable lines types - certain and maybe lines - certain lines should match the original result and maybe lines should match the original result assuming the analysed expression was not a const expr. Then (like you do now), reason the maybe lines using the certain ones. This will fix #953 and keep everyone happy.

Reply:
> You did not answer my question, what is the physical meaning (of emiting coverage for first and last string line, for empty lines, operator symbols, ...)?

My previous comments answer the question: the physical meaning (point [4] of my list) has no value when CC is unreliable

> So basically you say line coverage has little imformation value (qualitative value). This is not true, as line coverage is normally the primary value reported by tools like CodeCov.

Spread and adoption !== value

Line Code Coverage has little value because it can be easily cheated, both on purpose and unintentionally.
Line Code Coverage is good only for finding uncovered lines (red ones), but a single executed line can have many code branches and lead to many code paths, so seeing green doesn't mean your tests are good.

Only Mutation Testing can prove the tests are good. Line Code Coverage is still important though to ease the work and focus of MT on diffs.

> Also this PR has exactly opposite then little effect on results as 93.65% vs. 96.11% line coverage is huge change.

The amount of uncovered lines staid the same, so no change at all in my PoV.

> We both know quite well when/where xdebug emit coverage points and we know when it does not (always or sometimes due opcache). The result I expect from branch analysis PR is an update of the original executable analysis

Autoload runtimes are unpredictable, as well as opcache ones, so no one can predict where a CC driver is going emit a CC point for a particular test run.
This is way this PR cannot be an update of the current `9.2` implementation

Reply:
> My previous comments answer the question: the physical meaning (point [4] of my list) has no value when CC is unreliable

really? Please provide an example of unreliable, more precisely "unstable" line, ie. line that is sometimes present in xdebug as 1, sometimes as 0.

> Autoload runtimes are unpredictable, as well as opcache ones, so no one can predict where a CC driver is going emit a CC point for a particular test run.

AFAK this is not so much about CC drivers, but more about php-src, as php-src tries to compile const expr. code - but the executable lines are emit on quite well known places.

Reply:
> really? Please provide an example of unreliable, more precisely "unstable" line, ie. line that is sometimes present in xdebug as 1, sometimes as 0.

https://github.com/sebastianbergmann/php-code-coverage/pull/892

> AFAK this is not so much about CC drivers, but more about php-src, as php-src tries to compile const expr. code - but the executable lines are emit on quite well known places.

Exactly, and that's change nothing for us

Reply:
You have my support to show coverage support for unstable lines, we know exactly where the coverage point is emit (/wo opcache). This PR should do the analysis and if it can always reason the coverage point, it should add an executable line.